server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -201,8 +196,52 @@ remoteSchemaInterface schemaDoc defn@(G.InterfaceTypeDefinition description name
   -- TODO: another way to obtain 'possibleTypes' is to lookup all the object
   -- types in the schema document that claim to implement this interface.  We
   -- should have a check that expresses that that collection of objects is equal
-  -- to 'possibelTypes'.
-  pure $ void $ P.selectionSetInterface name description subFieldParsers objs
+  -- to 'possibleTypes'.
+  pure $ P.selectionSetInterface name description subFieldParsers objs <&>
+    (\objNameAndSelSets ->
+       -- This lambda function re-constructs a remote interface query.
+       -- We need to do this because the `SelectionSet`(s) that are
+       -- inputted to this function have the fragments (if any) flattened.
+       -- (Check `flattenSelectionSet` in 'Hasura.GraphQL.Parser.Collect' module)
+       -- This function makes a valid interface query by:
+       -- 1. Getting the common interface fields in all the selection sets
+       -- 2. Remove the common fields obtained in #1 from the selection sets
+       -- 3. Construct a selection field for every common interface field
+       -- 4. Construct inline fragments for non-common interface fields
+       --    using the result of #2 for every object
+       -- 5. Contruct the final selection set by combining #3 and #4
+       let selSetFields' = flip (fmap . fmap . fmap) objNameAndSelSets $
+                             \case
+                               G.SelectionField fld -> Just fld
+                               _                    -> Nothing
+           selSetFields = (fmap . fmap) catMaybes selSetFields'
```

Comment:
Can we rename these two variables to the role that they play in the algorithm, rather than essentially spelling out their type?

Reply:
Done!